File: modules/snoozepulse-audio/android/src/main/java/expo/modules/snoozepulseaudio/CaptureEngine.kt
package expo.modules.snoozepulseaudio

import android.content.Context
import android.content.Intent
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.media.AudioManager
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.os.Process
import android.util.Log
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.max

/**
 * Owns AudioRecord, the ring buffer, metering, snore detection, and snippet writes.
 * All PCM stays on this class — nothing is sent to JS except throttled metrics / events.
 *
 * Task 6.3 rewrites detection around the [SnoreClassifier] (ADR-21). The dB threshold
 * (`SNORE_MARGIN_DB`) is removed entirely per ADR-23. dB survives only as a display value
 * on `AudioLevelEvent` and as `SnoreEvent.peakDb` for legacy metric cards.
 */
internal class CaptureEngine(
  private val context: Context,
  private val classifier: SnoreClassifier?,
  private val emitLevel: (Map<String, Any?>) -> Unit,
  private val emitSnore: (Map<String, Any?>) -> Unit,
  private val emitInterruption: (Map<String, Any?>) -> Unit,
) {
  companion object {
    const val SAMPLE_RATE = 16_000
    private const val CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO
    private const val ENCODING = AudioFormat.ENCODING_PCM_16BIT
    /** Keep 5 s of PCM for snippets (acceptance: 5-second snippets). */
    private const val RING_SECONDS = 5
    private const val EMIT_INTERVAL_MS = 150L
    private const val CALIBRATION_MS = 3_000L

    // Task 6.3 hysteresis (ADR-21 / api-contracts.md).
    private const val SNORE_ENTER_CONFIDENCE = 0.55f
    private const val SNORE_EXIT_CONFIDENCE = 0.35f
    private const val SNORE_MIN_MS = 300L
    private const val SNORE_END_HANG_MS = 700L

    private const val CLASS_LABEL_SNORING = "snoring"
    private const val CLASS_LABEL_SNORT = "snort"
  }

  private val ringCapacity = SAMPLE_RATE * RING_SECONDS
  private val ring = PcmRingBuffer(ringCapacity)
  // Pre-allocated read / analysis windows — never new'd on the capture loop.
  private val readBuffer: ShortArray
  private val analysisWindow = ShortArray(SAMPLE_RATE / 10) // 100 ms worth
  private val snippetScratch = ShortArray(ringCapacity)
  private val waveform = WaveformWindow()
  private val patchScratch = ShortArray(waveform.patchSamples)

  private var audioRecord: AudioRecord? = null
  private var captureThread: HandlerThread? = null
  private var captureHandler: Handler? = null
  private var emitHandler: Handler? = null

  private val running = AtomicBoolean(false)
  private val paused = AtomicBoolean(false)

  @Volatile private var sessionId: String? = null
  @Volatile private var ambientBaselineDb: Double = 32.0

  private var snoreActive = false
  private var snoreStartedAt = 0L
  private var snorePeakDb = 0.0
  private var snorePeakConfidence = 0f
  private var snoreConfidenceSum = 0f
  private var snoreSnoringSum = 0f
  private var snoreSnortSum = 0f
  private var snoreClassifiedFrames = 0
  private var snoreLastAboveAt = 0L
  private var snoreEpisodeId: String? = null

  private var audioFocusChangeListener: AudioManager.OnAudioFocusChangeListener? = null

  init {
    val minBuf =
      AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL_CONFIG, ENCODING).coerceAtLeast(SAMPLE_RATE / 5)
    readBuffer = ShortArray(minBuf / 2) // shorts
  }

  fun isRunning(): Boolean = running.get()

  /**
   * Measure ambient floor for [CALIBRATION_MS]. Stops an orphaned capture first so a
   * force-killed session cannot block the next Start.
   */
  fun calibrate(): Map<String, Any> {
    if (running.get()) {
      stop()
    }
    val record = buildRecorder()
    if (record.state != AudioRecord.STATE_INITIALIZED) {
      record.release()
      throw IllegalStateException("Microphone is unavailable — close other apps using the mic and try again")
    }
    record.startRecording()
    val deadline = System.currentTimeMillis() + CALIBRATION_MS
    var sumDb = 0.0
    var samples = 0
    try {
      while (System.currentTimeMillis() < deadline) {
        val n = record.read(readBuffer, 0, readBuffer.size)
        if (n > 0) {
          val window = minOf(n, analysisWindow.size)
          System.arraycopy(readBuffer, 0, analysisWindow, 0, window)
          val rms = AudioDsp.rms(analysisWindow, window)
          sumDb += AudioDsp.rmsToDb(rms)
          samples += 1
        }
      }
    } finally {
      try {
        record.stop()
      } catch (_: Exception) {
      }
      record.release()
    }
    if (samples == 0) {
      throw IllegalStateException("No ambient samples captured — check microphone access and try again")
    }
    val baseline = sumDb / samples
    ambientBaselineDb = baseline
    val environment =
      when {
        baseline < 35.0 -> "quiet"
        baseline < 48.0 -> "moderate"
        else -> "noisy"
      }
    return mapOf(
      "baselineDb" to baseline,
      "sampledMs" to CALIBRATION_MS.toDouble(),
      "environment" to environment,
    )
  }

  fun start(sessionId: String, baselineDb: Double?) {
    if (!running.compareAndSet(false, true)) {
      throw IllegalStateException("Already recording")
    }
    this.sessionId = sessionId
    if (baselineDb != null) {
      ambientBaselineDb = baselineDb
    }
    paused.set(false)
    ring.clear()
    resetSnoreState()

    startForegroundService()
    requestAudioFocus()

    val record = buildRecorder()
    audioRecord = record
    record.startRecording()

    val thread = HandlerThread("SnoozePulseCapture", Process.THREAD_PRIORITY_AUDIO)
    thread.start()
    captureThread = thread
    captureHandler = Handler(thread.looper)
    emitHandler = Handler(context.mainLooper)

    captureHandler?.post(captureLoop)
    scheduleEmit()
  }

  fun stop() {
    if (!running.compareAndSet(true, false)) {
      return
    }
    finalizeSnoreIfNeeded(System.currentTimeMillis())
    captureHandler?.removeCallbacksAndMessages(null)
    emitHandler?.removeCallbacksAndMessages(null)
    try {
      audioRecord?.stop()
    } catch (_: Exception) {
    }
    audioRecord?.release()
    audioRecord = null
    captureThread?.quitSafely()
    captureThread = null
    captureHandler = null
    abandonAudioFocus()
    stopForegroundService()
    sessionId = null
  }

  fun pause() {
    if (!running.get() || !paused.compareAndSet(false, true)) {
      return
    }
    try {
      audioRecord?.stop()
    } catch (_: Exception) {
    }
    val id = sessionId ?: return
    emitInterruption(
      mapOf(
        "sessionId" to id,
        "timestamp" to System.currentTimeMillis().toDouble(),
        "reason" to "audioSessionSeized",
        "resumed" to false,
      ),
    )
  }

  fun resume() {
    if (!running.get() || !paused.compareAndSet(true, false)) {
      return
    }
    try {
      audioRecord?.startRecording()
    } catch (_: Exception) {
    }
    val id = sessionId ?: return
    emitInterruption(
      mapOf(
        "sessionId" to id,
        "timestamp" to System.currentTimeMillis().toDouble(),
        "reason" to "audioSessionSeized",
        "resumed" to true,
      ),
    )
  }

  private val captureLoop =
    object : Runnable {
      override fun run() {
        if (!running.get()) {
          return
        }
        val record = audioRecord
        if (record != null && !paused.get()) {
          val n = record.read(readBuffer, 0, readBuffer.size)
          if (n > 0) {
            ring.write(readBuffer, 0, n)
          }
        }
        captureHandler?.post(this)
      }
    }

  private fun scheduleEmit() {
    emitHandler?.post(
      object : Runnable {
        override fun run() {
          if (!running.get()) {
            return
          }
          if (!paused.get()) {
            emitMeterAndDetect()
          }
          emitHandler?.postDelayed(this, EMIT_INTERVAL_MS)
        }
      },
    )
  }

  private fun emitMeterAndDetect() {
    val id = sessionId ?: return
    val displayCount = ring.copyLatest(analysisWindow.size, analysisWindow)
    if (displayCount <= 0) {
      return
    }
    val rms = AudioDsp.rms(analysisWindow, displayCount)
    val peak = AudioDsp.peak(analysisWindow, displayCount)
    val db = AudioDsp.rmsToDb(rms)
    val peakDb = AudioDsp.rmsToDb(peak.toDouble())

    val classification = classifyIfReady()
    val confidence = classification?.combined ?: 0f
    val now = System.currentTimeMillis()

    // Detection is entirely classifier-driven now (ADR-23). Hysteresis state advances
    // regardless of dB; dB is kept only for display and for SnoreEvent.peakDb.
    updateSnoreState(now, peakDb, classification)

    emitLevel(
      mapOf(
        "sessionId" to id,
        "timestamp" to now.toDouble(),
        "decibel" to db,
        "rms" to rms / 32768.0,
        "snoreDetected" to snoreActive,
        "confidence" to confidence.toDouble(),
        // Task 6.4 will replace this with the rolling 60 s median (ADR-25). Emit 0.0
        // for now so downstream types are stable.
        "noiseFloorDb" to 0.0,
      ),
    )
  }

  /**
   * Run YAMNet on the newest 15,600 ring samples if we have that many yet. Returns null
   * during the first ~1 s of a session (ring not full enough) or when the classifier is
   * unavailable (warmup failed at module init).
   */
  private fun classifyIfReady(): SnoreClassifier.Classification? {
    val cls = classifier ?: return null
    if (ring.length < waveform.patchSamples) {
      return null
    }
    val read = ring.copyLatest(waveform.patchSamples, patchScratch)
    if (read != waveform.patchSamples) {
      return null
    }
    val patch = waveform.fill(patchScratch, waveform.patchSamples)
    return try {
      cls.classify(patch)
    } catch (error: Exception) {
      // Never let a classifier crash kill the capture loop.
      Log.w("CaptureEngine", "Classifier failure on capture path: ${error.message}")
      null
    }
  }

  private fun updateSnoreState(
    now: Long,
    peakDb: Double,
    classification: SnoreClassifier.Classification?,
  ) {
    val confidence = classification?.combined ?: 0f
    val above = confidence >= SNORE_ENTER_CONFIDENCE
    val below = confidence < SNORE_EXIT_CONFIDENCE

    if (above) {
      snoreLastAboveAt = now
      if (!snoreActive) {
        snoreActive = true
        snoreStartedAt = now
        snorePeakDb = peakDb
        snorePeakConfidence = confidence
        snoreConfidenceSum = 0f
        snoreSnoringSum = 0f
        snoreSnortSum = 0f
        snoreClassifiedFrames = 0
        snoreEpisodeId = UUID.randomUUID().toString()
      } else {
        if (peakDb > snorePeakDb) snorePeakDb = peakDb
        if (confidence > snorePeakConfidence) snorePeakConfidence = confidence
      }
    }

    if (snoreActive && classification != null) {
      snoreConfidenceSum += classification.combined
      snoreSnoringSum += classification.snoring
      snoreSnortSum += classification.snort
      snoreClassifiedFrames += 1
    }

    if (snoreActive && below && now - snoreLastAboveAt >= SNORE_END_HANG_MS) {
      finalizeSnoreIfNeeded(now)
    }
  }

  private fun finalizeSnoreIfNeeded(now: Long) {
    if (!snoreActive) {
      return
    }
    val duration = now - snoreStartedAt
    val id = sessionId
    val episodeId = snoreEpisodeId
    val frames = snoreClassifiedFrames
    val snoringTotal = snoreSnoringSum
    val snortTotal = snoreSnortSum
    val meanConfidence =
      if (frames > 0) (snoreConfidenceSum / frames.toFloat()).toDouble() else 0.0
    val classLabel = if (snoringTotal >= snortTotal) CLASS_LABEL_SNORING else CLASS_LABEL_SNORT
    snoreActive = false
    if (id == null || episodeId == null || duration < SNORE_MIN_MS) {
      resetSnoreState()
      return
    }
    val sampleCount = ring.copyLatest(ringCapacity, snippetScratch)
    val path =
      SnippetWriter.writeWav(
        context.filesDir,
        id,
        snoreStartedAt,
        SAMPLE_RATE,
        snippetScratch,
        sampleCount,
      )
    emitSnore(
      mapOf(
        "id" to episodeId,
        "sessionId" to id,
        "timestamp" to snoreStartedAt.toDouble(),
        "durationMs" to duration.toDouble(),
        "peakDb" to snorePeakDb,
        "confidence" to meanConfidence,
        "classLabel" to classLabel,
        // spectralPeakHz is emitted as null in Task 6.3; a future spectral-analysis stage
        // will populate it. Null is a legal value per docs/api-contracts.md.
        "spectralPeakHz" to null,
        "audioPath" to path,
      ),
    )
    resetSnoreState()
  }

  private fun resetSnoreState() {
    snoreActive = false
    snoreStartedAt = 0L
    snorePeakDb = 0.0
    snorePeakConfidence = 0f
    snoreConfidenceSum = 0f
    snoreSnoringSum = 0f
    snoreSnortSum = 0f
    snoreClassifiedFrames = 0
    snoreLastAboveAt = 0L
    snoreEpisodeId = null
  }

  private fun buildRecorder(): AudioRecord {
    val minBuf = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL_CONFIG, ENCODING)
    val bufferSize = max(minBuf, readBuffer.size * 2)
    return AudioRecord(
      MediaRecorder.AudioSource.MIC,
      SAMPLE_RATE,
      CHANNEL_CONFIG,
      ENCODING,
      bufferSize,
    )
  }

  private fun startForegroundService() {
    val intent = Intent(context, MicrophoneForegroundService::class.java)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      context.startForegroundService(intent)
    } else {
      context.startService(intent)
    }
  }

  private fun stopForegroundService() {
    context.stopService(Intent(context, MicrophoneForegroundService::class.java))
  }

  private fun requestAudioFocus() {
    val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    val listener =
      AudioManager.OnAudioFocusChangeListener { change ->
        when (change) {
          AudioManager.AUDIOFOCUS_LOSS,
          AudioManager.AUDIOFOCUS_LOSS_TRANSIENT,
          AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK,
          -> {
            if (running.get() && !paused.get()) {
              pause()
            }
          }
          AudioManager.AUDIOFOCUS_GAIN -> {
            if (running.get() && paused.get()) {
              resume()
            }
          }
        }
      }
    audioFocusChangeListener = listener
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      // Legacy request keeps API simple across OEM versions for V1.
      @Suppress("DEPRECATION")
      am.requestAudioFocus(listener, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN)
    } else {
      @Suppress("DEPRECATION")
      am.requestAudioFocus(listener, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN)
    }
  }

  private fun abandonAudioFocus() {
    val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    val listener = audioFocusChangeListener ?: return
    @Suppress("DEPRECATION")
    am.abandonAudioFocus(listener)
    audioFocusChangeListener = null
  }
}

File: modules/snoozepulse-audio/ios/CaptureEngine.swift
import AVFoundation
import Foundation

/**
 * AVAudioEngine capture + DSP. Mirrors Android CaptureEngine contracts so JS stays identical.
 * PCM never leaves this type — only throttled level / snore / interruption payloads.
 *
 * Input taps use the hardware float format, resample to 16 kHz Int16, then write a
 * pre-sized ring (no per-frame allocation of the ring itself).
 *
 * Task 6.3 rewrites detection around `SnoreClassifier` (ADR-21). The dB threshold
 * (`snoreMarginDb`) is removed entirely per ADR-23. dB survives only as a display value
 * on `AudioLevelEvent` and as `SnoreEvent.peakDb` for legacy metric cards.
 *
 * Task 6.4 (ADR-25) additionally:
 * - Retains `.measurement` mode (AGC disabled) and logs the session category / mode on
 *   start so we have parity with the Android capture-source log.
 * - Maintains a rolling 60 s median of display dB and publishes it as
 *   `AudioLevelEvent.noiseFloorDb`.
 * - Turns `calibrate()` into an immediate query over that rolling median (no 3 s wait).
 */
final class CaptureEngine {
  static let sampleRate = 16_000
  private static let ringSeconds = 5
  private static let emitIntervalMs: Double = 150

  // Task 6.3 hysteresis (ADR-21 / api-contracts.md).
  private static let snoreEnterConfidence: Float = 0.55
  private static let snoreExitConfidence: Float = 0.35
  private static let snoreMinMs: Int64 = 300
  private static let snoreEndHangMs: Int64 = 700

  private static let classLabelSnoring = "snoring"
  private static let classLabelSnort = "snort"

  // Task 6.4 noise-floor window (ADR-25). 60 s / emitIntervalMs ≈ 400 samples.
  private static let noiseFloorWindowMs: Double = 60_000
  private static let noiseFloorCapacity: Int = max(1, Int(noiseFloorWindowMs / emitIntervalMs))

  // Environment thresholds preserved from the M5 UX — see docs/api-contracts.md.
  private static let quietUpperDb: Double = 35
  private static let moderateUpperDb: Double = 48

  // Silence gate (Task battery-hardening §4). Skip YAMNet when the room is quiet.
  // Configurable in-place: change the constant, recompile, rerun the regression corpus.
  private static let silenceGateMarginDb: Double = 6.0
  private static let silenceGateMinIntervalMs: Int64 = 2_000

  // Debug telemetry cadence.
  private static let filterLogIntervalMs: Int64 = 60_000

  private let classifier: SnoreClassifier?

  private let ring: PcmRingBuffer
  private var analysisWindow: [Int16]
  private var snippetScratch: [Int16]
  private var convertScratch: [Int16]
  private let waveform = WaveformWindow()
  private var patchScratch: [Int16]
  private let noiseFloor = NoiseFloor(capacity: CaptureEngine.noiseFloorCapacity)

  private let engine = AVAudioEngine()
  private var isRunning = false
  private var isPaused = false
  private var sessionId: String?

  private var snoreActive = false
  private var snoreStartedAt: Int64 = 0
  private var snorePeakDb: Double = 0
  private var snorePeakConfidence: Float = 0
  private var snoreConfidenceSum: Float = 0
  private var snoreSnoringSum: Float = 0
  private var snoreSnortSum: Float = 0
  private var snoreClassifiedFrames: Int = 0
  private var snoreLastAboveAt: Int64 = 0
  private var snoreEpisodeId: String?

  // Silence-gate + quality-filter state. Reset in `resetSnoreState`.
  private var lastInferenceMonotonicMs: Int64 = 0
  private var totalWindows: Int = 0
  private var skippedByGate: Int = 0
  private var totalInferences: Int = 0
  private var lastFilterLogMs: Int64 = 0

  private var emitTimer: Timer?
  private var interruptionObserver: NSObjectProtocol?

  var onLevel: (([String: Any]) -> Void)?
  var onSnore: (([String: Any]) -> Void)?
  var onInterruption: (([String: Any]) -> Void)?

  init(classifier: SnoreClassifier? = nil) {
    self.classifier = classifier
    let capacity = Self.sampleRate * Self.ringSeconds
    ring = PcmRingBuffer(capacity: capacity)
    analysisWindow = [Int16](repeating: 0, count: Self.sampleRate / 10)
    snippetScratch = [Int16](repeating: 0, count: capacity)
    convertScratch = [Int16](repeating: 0, count: 4096)
    patchScratch = [Int16](repeating: 0, count: WaveformWindow.defaultPatchSamples)
  }

  func running() -> Bool { isRunning }

  /// Return the current rolling-median noise floor without blocking (ADR-25). The M5
  /// three-second sampling loop is gone; the noise floor is now maintained continuously
  /// during capture. On cold start (no capture has ever run) the floor is 0 and the
  /// environment is reported as "quiet" — the Home affordance stays as cosmetic feedback
  /// while the first minute of the rolling window fills.
  func calibrate() throws -> [String: Any] {
    let baseline = noiseFloor.current()
    return [
      "baselineDb": baseline,
      "sampledMs": 0.0,
      "environment": classifyEnvironment(baseline),
    ]
  }

  private func classifyEnvironment(_ baseline: Double) -> String {
    if baseline < Self.quietUpperDb { return "quiet" }
    if baseline < Self.moderateUpperDb { return "moderate" }
    return "noisy"
  }

  /// - Parameter baselineDb: Ignored under ADR-23/-25. The parameter is retained so the
  ///   Expo module callsite stays stable; the noise floor is now maintained by the rolling
  ///   median instead.
  func start(sessionId: String, baselineDb: Double?) throws {
    if isRunning {
      throw NSError(domain: "SnoozePulseAudio", code: 2, userInfo: [
        NSLocalizedDescriptionKey: "Already recording",
      ])
    }
    self.sessionId = sessionId
    isPaused = false
    ring.clear()
    noiseFloor.reset()
    resetSnoreState()
    try configureSession()
    observeInterruptions()

    let input = engine.inputNode
    let format = input.inputFormat(forBus: 0)
    input.installTap(onBus: 0, bufferSize: 1024, format: format) { [weak self] buffer, _ in
      guard let self, self.isRunning, !self.isPaused else { return }
      let count = self.floatBufferToInt16(buffer)
      guard count > 0 else { return }
      self.convertScratch.withUnsafeBufferPointer { ptr in
        guard let base = ptr.baseAddress else { return }
        self.ring.write(base, count: count)
      }
    }
    engine.prepare()
    try engine.start()
    isRunning = true
    let session = AVAudioSession.sharedInstance()
    NSLog(
      "CaptureEngine capture source granted: mode=%@ category=%@",
      session.mode.rawValue,
      session.category.rawValue
    )
    startEmitTimer()
  }

  func stop() {
    guard isRunning else { return }
    finalizeSnoreIfNeeded(now: Int64(Date().timeIntervalSince1970 * 1000))
    isRunning = false
    emitTimer?.invalidate()
    emitTimer = nil
    engine.inputNode.removeTap(onBus: 0)
    engine.stop()
    if let observer = interruptionObserver {
      NotificationCenter.default.removeObserver(observer)
      interruptionObserver = nil
    }
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    sessionId = nil
  }

  func pause() {
    guard isRunning, !isPaused else { return }
    isPaused = true
    engine.pause()
    guard let id = sessionId else { return }
    onInterruption?([
      "sessionId": id,
      "timestamp": Date().timeIntervalSince1970 * 1000,
      "reason": "audioSessionSeized",
      "resumed": false,
    ])
  }

  func resume() {
    guard isRunning, isPaused else { return }
    do {
      try engine.start()
      isPaused = false
      guard let id = sessionId else { return }
      onInterruption?([
        "sessionId": id,
        "timestamp": Date().timeIntervalSince1970 * 1000,
        "reason": "audioSessionSeized",
        "resumed": true,
      ])
    } catch {
      // Leave paused.
    }
  }

  private func configureSession() throws {
    let session = AVAudioSession.sharedInstance()
    try session.setCategory(
      .playAndRecord,
      mode: .measurement,
      options: [.mixWithOthers, .allowBluetooth, .allowBluetoothA2DP, .defaultToSpeaker]
    )
    // Prefer 16 kHz to match Android AudioRecord; hardware may still deliver 44.1/48 kHz.
    try? session.setPreferredSampleRate(Double(Self.sampleRate))
    try session.setActive(true, options: .notifyOthersOnDeactivation)
  }

  /// Convert input tap PCM to mono Int16 at `sampleRate` (resample when hardware ≠ 16 kHz).
  private func floatBufferToInt16(_ buffer: AVAudioPCMBuffer) -> Int {
    guard let channel = buffer.floatChannelData?[0] else { return 0 }
    let inFrames = Int(buffer.frameLength)
    guard inFrames > 0 else { return 0 }

    let inRate = buffer.format.sampleRate
    let outRate = Double(Self.sampleRate)
    let outCount: Int
    if abs(inRate - outRate) < 1.0 {
      outCount = inFrames
      ensureConvertCapacity(outCount)
      for i in 0..<inFrames {
        let clamped = max(-1.0, min(1.0, Double(channel[i])))
        convertScratch[i] = Int16(clamped * 32767.0)
      }
      return outCount
    }

    // Linear resample hardware rate → 16 kHz (same target as Android CaptureEngine).
    outCount = max(1, Int((Double(inFrames) * outRate / inRate).rounded(.down)))
    ensureConvertCapacity(outCount)
    let step = inRate / outRate
    for i in 0..<outCount {
      let srcPos = Double(i) * step
      let idx = min(inFrames - 1, Int(srcPos))
      let frac = srcPos - Double(idx)
      let s0 = Double(channel[idx])
      let s1 = Double(channel[min(inFrames - 1, idx + 1)])
      let sample = s0 + (s1 - s0) * frac
      let clamped = max(-1.0, min(1.0, sample))
      convertScratch[i] = Int16(clamped * 32767.0)
    }
    return outCount
  }

  private func ensureConvertCapacity(_ count: Int) {
    if count > convertScratch.count {
      convertScratch = [Int16](repeating: 0, count: count)
    }
  }

  private func observeInterruptions() {
    interruptionObserver = NotificationCenter.default.addObserver(
      forName: AVAudioSession.interruptionNotification,
      object: AVAudioSession.sharedInstance(),
      queue: .main
    ) { [weak self] notification in
      guard let self else { return }
      guard
        let info = notification.userInfo,
        let typeValue = info[AVAudioSessionInterruptionTypeKey] as? UInt,
        let type = AVAudioSession.InterruptionType(rawValue: typeValue)
      else { return }
      switch type {
      case .began:
        self.pause()
      case .ended:
        if let optionsValue = info[AVAudioSessionInterruptionOptionKey] as? UInt {
          let options = AVAudioSession.InterruptionOptions(rawValue: optionsValue)
          if options.contains(.shouldResume) {
            self.resume()
          }
        }
      @unknown default:
        break
      }
    }
  }

  private func startEmitTimer() {
    DispatchQueue.main.async {
      self.emitTimer = Timer.scheduledTimer(withTimeInterval: Self.emitIntervalMs / 1000.0, repeats: true) { [weak self] _ in
        self?.emitMeterAndDetect()
      }
    }
  }

  private func emitMeterAndDetect() {
    guard isRunning, !isPaused, let id = sessionId else { return }
    let displayCount = ring.copyLatest(count: analysisWindow.count, into: &analysisWindow)
    guard displayCount > 0 else { return }
    let rms = AudioDsp.rms(samples: analysisWindow, count: displayCount)
    let peak = AudioDsp.peak(samples: analysisWindow, count: displayCount)
    let db = AudioDsp.rmsToDb(rms)
    let peakDb = AudioDsp.rmsToDb(Double(peak))

    // Update the rolling 60 s noise-floor median before emitting so this tick's value
    // reflects the sample we just captured (ADR-25).
    noiseFloor.push(db)
    let noiseFloorDb = noiseFloor.current()

    let now = Int64(Date().timeIntervalSince1970 * 1000)
    totalWindows += 1
    let classification = classifyIfReady(nowMs: now, db: db, noiseFloorDb: noiseFloorDb)
    let confidence = classification?.combined ?? 0

    // Detection is entirely classifier-driven now (ADR-23). Hysteresis state advances
    // regardless of dB; dB is kept only for display and for SnoreEvent.peakDb.
    updateSnoreState(now: now, peakDb: peakDb, classification: classification)

    maybeLogFilterStats(now: now)

    onLevel?([
      "sessionId": id,
      "timestamp": Double(now),
      "decibel": db,
      "rms": rms / 32768.0,
      "snoreDetected": snoreActive,
      "confidence": Double(confidence),
      "noiseFloorDb": noiseFloorDb,
    ])
  }

  /// Run YAMNet on the newest 15,600 ring samples if we have that many yet.
  ///
  /// Silence gate (Task battery-hardening §4): once the rolling noise-floor window has warmed
  /// up, skip inference entirely when the current dB is below `noiseFloor + margin`, unless
  /// `silenceGateMinIntervalMs` has elapsed since the last inference (keep-alive so we never
  /// miss the leading edge of a soft snore that begins below floor + margin).
  ///
  /// Returns nil during the first ~1 s of a session (ring not full enough), when the
  /// classifier is unavailable (warmup failed at module init), or when the silence gate
  /// suppresses this window.
  private func classifyIfReady(nowMs: Int64, db: Double, noiseFloorDb: Double) -> SnoreClassification? {
    guard let cls = classifier else { return nil }
    guard ring.length >= waveform.patchSamples else { return nil }

    // Silence gate: only engage after the noise-floor rolling window is warm. On cold start
    // `noiseFloorDb == 0` — do not compare a real dB against 0 or every window would run.
    let noiseFloorWarm = noiseFloorDb > 0
    if noiseFloorWarm {
      let elevated = db >= noiseFloorDb + Self.silenceGateMarginDb
      let keepAlive = nowMs - lastInferenceMonotonicMs >= Self.silenceGateMinIntervalMs
      if !elevated && !keepAlive {
        skippedByGate += 1
        return nil
      }
    }

    let read = ring.copyLatest(count: waveform.patchSamples, into: &patchScratch)
    guard read == waveform.patchSamples else { return nil }
    let patch = waveform.fill(patchScratch)
    do {
      let classification = try cls.classify(patch: patch)
      lastInferenceMonotonicMs = nowMs
      totalInferences += 1
      return classification
    } catch {
      NSLog("Classifier failure on capture path: %@", String(describing: error))
      return nil
    }
  }

  private func maybeLogFilterStats(now: Int64) {
    if lastFilterLogMs == 0 {
      lastFilterLogMs = now
      return
    }
    if now - lastFilterLogMs < Self.filterLogIntervalMs { return }
    NSLog(
      "SnoozePulse gate: windows=%d skipped=%d inferences=%d",
      totalWindows,
      skippedByGate,
      totalInferences
    )
    lastFilterLogMs = now
    totalWindows = 0
    skippedByGate = 0
    totalInferences = 0
  }

  private func updateSnoreState(now: Int64, peakDb: Double, classification: SnoreClassification?) {
    let confidence = classification?.combined ?? 0
    let above = confidence >= Self.snoreEnterConfidence
    let below = confidence < Self.snoreExitConfidence

    if above {
      snoreLastAboveAt = now
      if !snoreActive {
        snoreActive = true
        snoreStartedAt = now
        snorePeakDb = peakDb
        snorePeakConfidence = confidence
        snoreConfidenceSum = 0
        snoreSnoringSum = 0
        snoreSnortSum = 0
        snoreClassifiedFrames = 0
        snoreEpisodeId = UUID().uuidString
      } else {
        if peakDb > snorePeakDb { snorePeakDb = peakDb }
        if confidence > snorePeakConfidence { snorePeakConfidence = confidence }
      }
    }

    if snoreActive, let classification {
      snoreConfidenceSum += classification.combined
      snoreSnoringSum += classification.snoring
      snoreSnortSum += classification.snort
      snoreClassifiedFrames += 1
    }

    if snoreActive, below, now - snoreLastAboveAt >= Self.snoreEndHangMs {
      finalizeSnoreIfNeeded(now: now)
    }
  }

  private func finalizeSnoreIfNeeded(now: Int64) {
    guard snoreActive else { return }
    let duration = now - snoreStartedAt
    let id = sessionId
    let episodeId = snoreEpisodeId
    let frames = snoreClassifiedFrames
    let snoringTotal = snoreSnoringSum
    let snortTotal = snoreSnortSum
    let meanConfidence: Double = frames > 0
      ? Double(snoreConfidenceSum) / Double(frames)
      : 0
    let classLabel = snoringTotal >= snortTotal ? Self.classLabelSnoring : Self.classLabelSnort
    snoreActive = false
    guard let id, let episodeId, duration >= Self.snoreMinMs else {
      resetSnoreState()
      return
    }
    let sampleCount = ring.copyLatest(count: snippetScratch.count, into: &snippetScratch)
    let path = SnippetWriter.writeWav(
      sessionId: id,
      timestampMs: snoreStartedAt,
      sampleRate: Self.sampleRate,
      samples: snippetScratch,
      sampleCount: sampleCount
    )
    onSnore?([
      "id": episodeId,
      "sessionId": id,
      "timestamp": Double(snoreStartedAt),
      "durationMs": Double(duration),
      "peakDb": snorePeakDb,
      "confidence": meanConfidence,
      "classLabel": classLabel,
      // spectralPeakHz is emitted as null in Task 6.3; a future spectral-analysis stage
      // will populate it. Null is a legal value per docs/api-contracts.md.
      "spectralPeakHz": NSNull(),
      "audioPath": path as Any,
    ])
    resetSnoreState()
  }

  private func resetSnoreState() {
    snoreActive = false
    snoreStartedAt = 0
    snorePeakDb = 0
    snorePeakConfidence = 0
    snoreConfidenceSum = 0
    snoreSnoringSum = 0
    snoreSnortSum = 0
    snoreClassifiedFrames = 0
    snoreLastAboveAt = 0
    snoreEpisodeId = nil
    lastInferenceMonotonicMs = 0
    totalWindows = 0
    skippedByGate = 0
    totalInferences = 0
    lastFilterLogMs = 0
  }

  /// Rolling median of display dB over the last `capacity` emit ticks (ADR-25). Fixed
  /// pre-allocated storage, no allocations after construction. The median-per-query is an
  /// `O(n log n)` sort on ≈ 400 doubles — well under one millisecond and only runs on the
  /// emit timer, once per ~150 ms.
  private final class NoiseFloor {
    private let capacity: Int
    private var ring: [Double]
    private var scratch: [Double]
    private var writeIndex = 0
    private var filled = 0
    private let lock = NSLock()

    init(capacity: Int) {
      self.capacity = capacity
      ring = [Double](repeating: 0, count: capacity)
      scratch = [Double](repeating: 0, count: capacity)
    }

    func push(_ db: Double) {
      lock.lock()
      defer { lock.unlock() }
      ring[writeIndex] = db
      writeIndex = (writeIndex + 1) % capacity
      if filled < capacity { filled += 1 }
    }

    func current() -> Double {
      lock.lock()
      defer { lock.unlock() }
      guard filled > 0 else { return 0 }
      for i in 0..<filled { scratch[i] = ring[i] }
      // Sort only the filled prefix; the tail is stale storage we don't consult.
      scratch[0..<filled].sort()
      return scratch[filled / 2]
    }

    func reset() {
      lock.lock()
      defer { lock.unlock() }
      writeIndex = 0
      filled = 0
    }
  }
}
